File: LibraryBuildProject/helubottombuttonsheet/src/main/java/cz/helu/helubottombuttonsheet/HeluBottomButtonSheet.java
package cz.helu.helubottombuttonsheet;


import android.annotation.SuppressLint;
import android.app.ActionBar;
import android.app.Dialog;
import android.content.Context;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.support.annotation.ColorInt;
import android.support.annotation.ColorRes;
import android.support.annotation.DimenRes;
import android.support.annotation.NonNull;
import android.support.annotation.StringRes;
import android.support.design.widget.BottomSheetDialogFragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.content.ContextCompat;
import android.util.DisplayMetrics;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import cz.helu.helubottombuttonsheet.entity.BaseSheetItem;
import cz.helu.helubottombuttonsheet.entity.CustomViewSheetItem;
import cz.helu.helubottombuttonsheet.entity.DividerSheetItem;
import cz.helu.helubottombuttonsheet.entity.TextSheetItem;
import cz.helu.helubottombuttonsheet.utility.DrawableUtility;


@SuppressLint("ValidFragment")
@SuppressWarnings("unused")
public class HeluBottomButtonSheet extends BottomSheetDialogFragment
{
	private int titleItemHeight;
	private int itemHeight;
	private int itemImageSize;
	private int spacingHorizontal;
	private int paddingVertical;
	private int itemTouchFeedbackColor;
	private int sheetBackgroundColor;
	private int sheetTitleColor;
	private int itemTextColor;
	private int dividerColor;
	private String title;
	private List<BaseSheetItem> mItemList = new ArrayList<>();


	@SuppressLint("ValidFragment")
	public HeluBottomButtonSheet(Builder builder)
	{
		this.titleItemHeight = builder.titleItemHeight;
		this.itemHeight = builder.itemHeight;
		this.itemImageSize = builder.itemImageSize;
		this.spacingHorizontal = builder.spacingHorizontal;
		this.paddingVertical = builder.paddingVertical;
		this.itemTouchFeedbackColor = builder.itemTouchFeedbackColor;
		this.sheetBackgroundColor = builder.sheetBackgroundColor;
		this.sheetTitleColor = builder.sheetTitleColor;
		this.itemTextColor = builder.itemTextColor;
		this.dividerColor = builder.dividerColor;
		this.title = builder.title;
	}


	@SuppressLint("RestrictedApi")
	@Override
	public void setupDialog(final Dialog dialog, int style)
	{
		super.setupDialog(dialog, style);

		// Setup View
		LinearLayout contentLayout = new LinearLayout(getContext());
		contentLayout.setOrientation(LinearLayout.VERTICAL);
		contentLayout.setBackgroundColor(sheetBackgroundColor);

		if(!title.isEmpty())
		{
			contentLayout.setPadding(0, 0, 0, paddingVertical);
			contentLayout.addView(createTitleView());
		}
		else
		{
			contentLayout.setPadding(0, paddingVertical, 0, paddingVertical);
		}

		for(BaseSheetItem entity : mItemList)
		{
			if(entity instanceof DividerSheetItem)
			{
				contentLayout.addView(createDividerItemView());
			}
			else if(entity instanceof TextSheetItem)
			{
				contentLayout.addView(createTextItemView((TextSheetItem) entity));
			}
			else if(entity instanceof CustomViewSheetItem)
			{
				if(((CustomViewSheetItem) entity).customView.getParent() == null)
					contentLayout.addView(((CustomViewSheetItem) entity).customView);
			}
		}

		dialog.setContentView(contentLayout);
	}


	public void show(FragmentManager manager)
	{
		show(manager, getTag());
	}


	@SuppressWarnings("unused")
	public void addButton(@NonNull String text, @NonNull View.OnClickListener clickListener)
	{
		mItemList.add(new TextSheetItem(text, clickListener));
	}


	@SuppressWarnings("unused")
	public void addButton(@NonNull Drawable drawable, @NonNull String text, @NonNull View.OnClickListener clickListener)
	{
		mItemList.add(new TextSheetItem(drawable, text, clickListener));
	}


	@SuppressWarnings("unused")
	public void addButton(int drawableResource, @NonNull String text, @NonNull View.OnClickListener clickListener)
	{
		mItemList.add(new TextSheetItem(drawableResource, text, clickListener));
	}


	@SuppressWarnings("unused")
	public void addDivider()
	{
		mItemList.add(new DividerSheetItem());
	}


	@SuppressWarnings("unused")
	public void addCustomView(@NonNull View customView)
	{
		mItemList.add(new CustomViewSheetItem(customView));
	}


	@SuppressWarnings("unused")
	public void addCustomView(@NonNull View customView, @NonNull View.OnClickListener clickListener)
	{
		mItemList.add(new CustomViewSheetItem(customView, clickListener));
	}


	private int convertDpToPx(int dp)
	{
		return (getContext() != null) ? (Math.round(dp * getContext().getResources().getDisplayMetrics().xdpi / DisplayMetrics.DENSITY_DEFAULT)) : 0;
	}


	private View createTitleView()
	{
		LinearLayout.LayoutParams textViewParams = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, titleItemHeight);

		TextView textView = new TextView(getContext());
		textView.setText(title);
		textView.setTextColor(sheetTitleColor);
		textView.setLayoutParams(textViewParams);
		textView.setGravity(Gravity.CENTER_VERTICAL);
		textView.setPadding(spacingHorizontal, 0, spacingHorizontal, 0);

		return textView;
	}


	private View createDividerItemView()
	{
		int verticalContentPadding = convertDpToPx(Builder.DEFAULT_CONTENT_VERTICAL_SPACING);
		LinearLayout.LayoutParams dividerParams = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, convertDpToPx(1));
		dividerParams.setMargins(0, verticalContentPadding, 0, verticalContentPadding);

		View divider = new View(getContext());
		divider.setLayoutParams(dividerParams);
		divider.setBackgroundColor(dividerColor);
		return divider;
	}


	private LinearLayout createTextItemView(TextSheetItem entity)
	{
		Drawable drawable = DrawableUtility.getAdaptiveRippleDrawable(sheetBackgroundColor, itemTouchFeedbackColor);
		LinearLayout item = new LinearLayout(getContext());
		LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(ActionBar.LayoutParams.MATCH_PARENT, itemHeight);

		item.setLayoutParams(params);
		item.setGravity(Gravity.CENTER_VERTICAL);
		item.setOrientation(LinearLayout.HORIZONTAL);
		item.setPadding(spacingHorizontal, 0, spacingHorizontal, 0);
		item.setOnClickListener(entity.clickListener);
		item.setBackground(drawable);

		if(entity.drawable != null || entity.drawableResource != -1)
		{
			ImageView imageView = new ImageView(getContext());
			LinearLayout.LayoutParams imageViewParams = new LinearLayout.LayoutParams(itemImageSize, itemImageSize);
			imageViewParams.setMargins(0, 0, spacingHorizontal, 0);
			imageView.setLayoutParams(imageViewParams);
			if(entity.drawable != null)
				imageView.setImageDrawable(entity.drawable);
			else
				imageView.setImageResource(entity.drawableResource);

			item.addView(imageView);
		}

		TextView textView = new TextView(getContext());
		textView.setText(entity.text);
		textView.setTextColor(itemTextColor);

		item.addView(textView);

		return item;
	}


	public static class Builder
	{
		static final int DEFAULT_TITLE_ITEM_HEIGHT = 56;
		static final int DEFAULT_ITEM_HEIGHT = 48;
		static final int DEFAULT_ITEM_IMAGE_SIZE = 24;
		static final int DEFAULT_HORIZONTAL_SPACING = 16;
		static final int DEFAULT_CONTENT_VERTICAL_SPACING = 8;

		private Context mContext;
		private int titleItemHeight;
		private int itemHeight;
		private int itemImageSize;
		private int spacingHorizontal;
		private int paddingVertical;
		private String title = "";
		@ColorInt
		private int itemTouchFeedbackColor = Color.LTGRAY;
		@ColorInt
		private int sheetBackgroundColor = Color.WHITE;
		@ColorInt
		private int sheetTitleColor = Color.GRAY;
		@ColorInt
		private int itemTextColor = Color.BLACK;
		@ColorInt
		private int dividerColor = Color.LTGRAY;


		public Builder(Context context)
		{
			mContext = context;

			titleItemHeight = convertDpToPx(DEFAULT_TITLE_ITEM_HEIGHT);
			itemHeight = convertDpToPx(DEFAULT_ITEM_HEIGHT);
			itemImageSize = convertDpToPx(DEFAULT_ITEM_IMAGE_SIZE);
			spacingHorizontal = convertDpToPx(DEFAULT_HORIZONTAL_SPACING);
			paddingVertical = convertDpToPx(DEFAULT_CONTENT_VERTICAL_SPACING);
		}


		@SuppressWarnings("unused")
		public Builder withTitleRes(@StringRes int titleResId)
		{
			this.title = mContext.getResources().getString(titleResId);
			return this;
		}


		@SuppressWarnings("unused")
		public Builder withTitle(String title)
		{
			this.title = title;
			return this;
		}


		@SuppressWarnings("unused")
		@SuppressLint("ResourceType")
		public Builder withTitleItemHeightRes(@DimenRes int dimension)
		{
			this.titleItemHeight = mContext.getResources().getDimensionPixelSize(dimension);
			return this;
		}


		@SuppressWarnings("unused")
		@SuppressLint("ResourceType")
		public Builder withTitleItemHeight(int pixelSize)
		{
			this.titleItemHeight = pixelSize;
			return this;
		}


		@SuppressWarnings("unused")
		@SuppressLint("ResourceType")
		public Builder withItemHeightRes(@DimenRes int dimension)
		{
			this.itemHeight = mContext.getResources().getDimensionPixelSize(dimension);
			return this;
		}


		@SuppressWarnings("unused")
		@SuppressLint("ResourceType")
		public Builder withItemHeight(int pixelSize)
		{
			this.itemHeight = pixelSize;
			return this;
		}


		@SuppressWarnings("unused")
		@SuppressLint("ResourceType")
		public Builder withImageSizeRes(@DimenRes int dimension)
		{
			this.itemImageSize = mContext.getResources().getDimensionPixelSize(dimension);
			return this;
		}


		@SuppressWarnings("unused")
		@SuppressLint("ResourceType")
		public Builder withImageSize(@DimenRes int pixelSize)
		{
			this.itemImageSize = pixelSize;
			return this;
		}


		@SuppressWarnings("unused")
		@SuppressLint("ResourceType")
		public Builder withHorizontalSpacingRes(@DimenRes int dimension)
		{
			this.spacingHorizontal = mContext.getResources().getDimensionPixelSize(dimension);
			return this;
		}


		@SuppressWarnings("unused")
		@SuppressLint("ResourceType")
		public Builder withHorizontalSpacing(int pixelSize)
		{
			this.spacingHorizontal = pixelSize;
			return this;
		}


		@SuppressWarnings("unused")
		@SuppressLint("ResourceType")
		public Builder withVerticalPaddingRes(@DimenRes int dimension)
		{
			this.paddingVertical = mContext.getResources().getDimensionPixelSize(dimension);
			return this;
		}


		@SuppressWarnings("unused")
		@SuppressLint("ResourceType")
		public Builder withVerticalPadding(int pixelSize)
		{
			this.paddingVertical = pixelSize;
			return this;
		}


		@SuppressWarnings("unused")
		public Builder withBackgroundColorRes(@ColorRes int color)
		{
			this.sheetBackgroundColor = ContextCompat.getColor(mContext, color);
			return this;
		}


		@SuppressWarnings("unused")
		public Builder withBackgroundColor(int color)
		{
			this.sheetBackgroundColor = color;
			return this;
		}


		@SuppressWarnings("unused")
		public Builder withItemTouchFeedbackColorRes(@ColorRes int color)
		{
			this.itemTouchFeedbackColor = ContextCompat.getColor(mContext, color);
			return this;
		}


		@SuppressWarnings("unused")
		public Builder withItemTouchFeedbackColor(int color)
		{
			this.itemTouchFeedbackColor = color;
			return this;
		}


		@SuppressWarnings("unused")
		public Builder withTitleColorRes(@ColorRes int color)
		{
			this.sheetTitleColor = ContextCompat.getColor(mContext, color);
			return this;
		}


		@SuppressWarnings("unused")
		public Builder withTitleColor(int color)
		{
			this.sheetTitleColor = color;
			return this;
		}


		@SuppressWarnings("unused")
		public Builder withItemTextColorRes(@ColorRes int color)
		{
			this.itemTextColor = ContextCompat.getColor(mContext, color);
			return this;
		}


		@SuppressWarnings("unused")
		public Builder withItemTextColor(int color)
		{
			this.itemTextColor = color;
			return this;
		}


		@SuppressWarnings("unused")
		public Builder withDividerColorRes(@ColorRes int color)
		{
			this.dividerColor = ContextCompat.getColor(mContext, color);
			return this;
		}


		@SuppressWarnings("unused")
		public Builder withDividerColor(int color)
		{
			this.dividerColor = color;
			return this;
		}


		@SuppressWarnings("unused")
		public HeluBottomButtonSheet build()
		{
			return new HeluBottomButtonSheet(this);
		}


		int convertDpToPx(int dp)
		{
			return Math.round(dp * mContext.getResources().getDisplayMetrics().xdpi / DisplayMetrics.DENSITY_DEFAULT);
		}
	}
}

File: LibraryBuildProject/helubottombuttonsheet/src/main/java/cz/helu/helubottombuttonsheet/HeluBottomButtonSheet.kt
package cz.helu.helubottombuttonsheet


import android.annotation.SuppressLint
import android.app.ActionBar
import android.app.Dialog
import android.content.Context
import android.graphics.Color
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.support.annotation.ColorInt
import android.support.annotation.ColorRes
import android.support.annotation.DimenRes
import android.support.annotation.StringRes
import android.support.design.widget.BottomSheetDialogFragment
import android.support.v4.app.FragmentManager
import android.support.v4.content.ContextCompat
import android.util.DisplayMetrics
import android.util.TypedValue
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import cz.helu.helubottombuttonsheet.entity.BaseSheetItem
import cz.helu.helubottombuttonsheet.entity.CustomViewSheetItem
import cz.helu.helubottombuttonsheet.entity.DividerSheetItem
import cz.helu.helubottombuttonsheet.entity.TextSheetItem
import cz.helu.helubottombuttonsheet.utility.DrawableUtility
import java.util.*


@Suppress("unused")
class HeluBottomButtonSheet : BottomSheetDialogFragment() {
	private var titleItemHeight: Int = 0
	private var itemHeight: Int = 0
	private var itemImageSize: Int = 0
	private var titleTextSize: Int = 0
	private var itemTextSize: Int = 0
	private var spacingHorizontal: Int = 0
	private var paddingVertical: Int = 0
	private var itemTouchFeedbackColor: Int = 0
	private var sheetBackgroundColor: Int = 0
	private var sheetTitleColor: Int = 0
	private var itemTextColor: Int = 0
	private var dividerColor: Int = 0
	private var title: String? = null
	private val itemList = ArrayList<BaseSheetItem>()
	private var contentLayout: LinearLayout? = null
	private var savedInstanceStateDone: Boolean = false


	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)

		val arguments = arguments
		if (arguments != null) {
			this.titleItemHeight = arguments.getInt(Builder.ARGUMENTS_TITLE_ITEM_HEIGHT)
			this.itemHeight = arguments.getInt(Builder.ARGUMENTS_ITEM_HEIGHT)
			this.itemImageSize = arguments.getInt(Builder.ARGUMENTS_ITEM_IMAGE_SIZE)
			this.titleTextSize = arguments.getInt(Builder.ARGUMENTS_TITLE_TEXT_SIZE)
			this.itemTextSize = arguments.getInt(Builder.ARGUMENTS_ITEM_TEXT_SIZE)
			this.spacingHorizontal = arguments.getInt(Builder.ARGUMENTS_SPACING_HORIZONTAL)
			this.paddingVertical = arguments.getInt(Builder.ARGUMENTS_PADDING_VERTICAL)
			this.itemTouchFeedbackColor = arguments.getInt(Builder.ARGUMENTS_ITEM_TOUCH_FEEDBACK_COLOR)
			this.sheetBackgroundColor = arguments.getInt(Builder.ARGUMENTS_SHEET_BACKGROUND_COLOR)
			this.sheetTitleColor = arguments.getInt(Builder.ARGUMENTS_SHEET_TITLE_COLOR)
			this.itemTextColor = arguments.getInt(Builder.ARGUMENTS_ITEM_TEXT_COLOR)
			this.dividerColor = arguments.getInt(Builder.ARGUMENTS_DIVIDER_COLOR)
			this.title = arguments.getString(Builder.ARGUMENTS_TITLE)
		}
	}


	override fun onResume() {
		super.onResume()
		savedInstanceStateDone = false
	}


	override fun onStart() {
		super.onStart()
		savedInstanceStateDone = false
	}


	override fun onSaveInstanceState(outState: Bundle) {
		super.onSaveInstanceState(outState)
		savedInstanceStateDone = true
	}


	@SuppressLint("RestrictedApi")
	override fun setupDialog(dialog: Dialog, style: Int) {
		super.setupDialog(dialog, style)

		// Setup View
		contentLayout = LinearLayout(context)
		contentLayout?.orientation = LinearLayout.VERTICAL
		contentLayout?.setBackgroundColor(sheetBackgroundColor)

		inflateViews()

		dialog.setContentView(contentLayout)
	}


	fun show(manager: FragmentManager?) {
		if (!savedInstanceStateDone && manager != null)
			show(manager, tag)
	}


	fun addButton(button: TextSheetItem): TextSheetItem {
		itemList.add(button)
		return button
	}


	fun addButton(text: String, clickListener: View.OnClickListener): TextSheetItem {
		val item = TextSheetItem(text, clickListener)
		itemList.add(item)
		return item
	}


	fun addButton(drawable: Drawable, text: String, clickListener: View.OnClickListener): TextSheetItem {
		val item = TextSheetItem(drawable, text, clickListener)
		itemList.add(item)
		return item
	}


	fun addButton(drawableResource: Int, text: String, clickListener: View.OnClickListener): TextSheetItem {
		val item = TextSheetItem(drawableResource, text, clickListener)
		itemList.add(item)
		return item
	}


	fun addDivider(): DividerSheetItem {
		val item = DividerSheetItem()
		itemList.add(item)
		return item
	}


	fun addCustomView(customView: View): CustomViewSheetItem {
		val item = CustomViewSheetItem(customView)
		itemList.add(item)
		return item
	}


	fun addCustomView(customView: View, clickListener: View.OnClickListener): CustomViewSheetItem {
		val item = CustomViewSheetItem(customView, clickListener)
		itemList.add(item)
		return item
	}


	fun getItem(position: Int): BaseSheetItem {
		return itemList[position]
	}


	fun invalidate() {
		contentLayout?.removeAllViews()
		inflateViews()
	}


	private fun convertDpToPx(dp: Int): Int {
		return Math.round(dp * (context?.resources?.displayMetrics?.xdpi ?: 0f) / DisplayMetrics.DENSITY_DEFAULT)
	}


	private fun createTitleView(): View {
		val textViewParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, titleItemHeight)

		val textView = TextView(context)
		textView.text = title
		textView.setTextColor(sheetTitleColor)
		textView.layoutParams = textViewParams
		textView.gravity = Gravity.CENTER_VERTICAL
		textView.setTextSize(TypedValue.COMPLEX_UNIT_PX, titleTextSize.toFloat())
		textView.setPadding(spacingHorizontal, 0, spacingHorizontal, 0)

		return textView
	}


	private fun createDividerItemView(): View {
		val verticalContentPadding = convertDpToPx(Builder.DEFAULT_CONTENT_VERTICAL_SPACING)
		val dividerParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, convertDpToPx(1))
		dividerParams.setMargins(0, verticalContentPadding, 0, verticalContentPadding)

		val divider = View(context)
		divider.layoutParams = dividerParams
		divider.setBackgroundColor(dividerColor)
		return divider
	}


	private fun createTextItemView(entity: TextSheetItem): LinearLayout {
		val drawable = DrawableUtility.getAdaptiveRippleDrawable(sheetBackgroundColor, itemTouchFeedbackColor)
		val item = LinearLayout(context)
		val params = LinearLayout.LayoutParams(ActionBar.LayoutParams.MATCH_PARENT, itemHeight)

		item.layoutParams = params
		item.gravity = Gravity.CENTER_VERTICAL
		item.orientation = LinearLayout.HORIZONTAL
		item.setPadding(spacingHorizontal, 0, spacingHorizontal, 0)
		item.setOnClickListener(entity.clickListener)
		item.background = drawable

		if (entity.drawable != null || entity.drawableResource != -1) {
			val imageView = ImageView(context)
			val imageViewParams = LinearLayout.LayoutParams(itemImageSize, itemImageSize)
			imageViewParams.setMargins(0, 0, spacingHorizontal, 0)
			imageView.layoutParams = imageViewParams
			if (entity.drawable != null)
				imageView.setImageDrawable(entity.drawable)
			else
				imageView.setImageResource(entity.drawableResource)

			item.addView(imageView)
		}

		val textView = TextView(context)
		textView.text = entity.text
		textView.setTextColor(itemTextColor)
		textView.setTextSize(TypedValue.COMPLEX_UNIT_PX, itemTextSize.toFloat())

		item.addView(textView)

		return item
	}


	private fun inflateViews() {
		if (title?.isNotEmpty() == true) {
			contentLayout?.setPadding(0, 0, 0, paddingVertical)
			contentLayout?.addView(createTitleView())
		} else {
			contentLayout?.setPadding(0, paddingVertical, 0, paddingVertical)
		}

		for (entity in itemList) {
			if (entity is DividerSheetItem) {
				contentLayout?.addView(createDividerItemView())
			} else if (entity is TextSheetItem) {
				contentLayout?.addView(createTextItemView(entity))
			} else if (entity is CustomViewSheetItem) {
				if (entity.customView.parent == null)
					contentLayout?.addView(entity.customView)
			}
		}
	}


	class Builder(private val context: Context) {
		companion object {
			internal const val ARGUMENTS_TITLE_ITEM_HEIGHT = "ARGUMENTS_TITLE_ITEM_HEIGHT"
			internal const val ARGUMENTS_ITEM_HEIGHT = "ARGUMENTS_ITEM_HEIGHT"
			internal const val ARGUMENTS_ITEM_IMAGE_SIZE = "ARGUMENTS_ITEM_IMAGE_SIZE"
			internal const val ARGUMENTS_TITLE_TEXT_SIZE = "ARGUMENTS_TITLE_TEXT_SIZE"
			internal const val ARGUMENTS_ITEM_TEXT_SIZE = "ARGUMENTS_ITEM_TEXT_SIZE"
			internal const val ARGUMENTS_SPACING_HORIZONTAL = "ARGUMENTS_SPACING_HORIZONTAL"
			internal const val ARGUMENTS_PADDING_VERTICAL = "ARGUMENTS_PADDING_VERTICAL"
			internal const val ARGUMENTS_ITEM_TOUCH_FEEDBACK_COLOR = "ARGUMENTS_ITEM_TOUCH_FEEDBACK_COLOR"
			internal const val ARGUMENTS_SHEET_BACKGROUND_COLOR = "ARGUMENTS_SHEET_BACKGROUND_COLOR"
			internal const val ARGUMENTS_SHEET_TITLE_COLOR = "ARGUMENTS_SHEET_TITLE_COLOR"
			internal const val ARGUMENTS_ITEM_TEXT_COLOR = "ARGUMENTS_ITEM_TEXT_COLOR"
			internal const val ARGUMENTS_DIVIDER_COLOR = "ARGUMENTS_DIVIDER_COLOR"
			internal const val ARGUMENTS_TITLE = "ARGUMENTS_TITLE"

			internal const val DEFAULT_TITLE_ITEM_HEIGHT = 56
			internal const val DEFAULT_ITEM_HEIGHT = 48
			internal const val DEFAULT_ITEM_IMAGE_SIZE = 24
			internal const val DEFAULT_TITLE_TEXT_SIZE = 14
			internal const val DEFAULT_ITEM_TEXT_SIZE = 14
			internal const val DEFAULT_HORIZONTAL_SPACING = 16
			internal const val DEFAULT_CONTENT_VERTICAL_SPACING = 8
		}

		private var titleItemHeight: Int = 0
		private var itemHeight: Int = 0
		private var itemImageSize: Int = 0
		private var titleTextSize: Int = 0
		private var itemTextSize: Int = 0
		private var spacingHorizontal: Int = 0
		private var paddingVertical: Int = 0
		private var title = ""
		@ColorInt
		private var itemTouchFeedbackColor = Color.LTGRAY
		@ColorInt
		private var sheetBackgroundColor = Color.parseColor("#FAFAFA")
		@ColorInt
		private var sheetTitleColor = Color.GRAY
		@ColorInt
		private var itemTextColor = Color.BLACK
		@ColorInt
		private var dividerColor = Color.LTGRAY


		init {

			titleItemHeight = convertDpToPx(DEFAULT_TITLE_ITEM_HEIGHT)
			itemHeight = convertDpToPx(DEFAULT_ITEM_HEIGHT)
			itemImageSize = convertDpToPx(DEFAULT_ITEM_IMAGE_SIZE)
			titleTextSize = convertDpToPx(DEFAULT_TITLE_TEXT_SIZE)
			itemTextSize = convertDpToPx(DEFAULT_ITEM_TEXT_SIZE)
			spacingHorizontal = convertDpToPx(DEFAULT_HORIZONTAL_SPACING)
			paddingVertical = convertDpToPx(DEFAULT_CONTENT_VERTICAL_SPACING)
		}


		fun withTitleRes(@StringRes titleResId: Int): Builder {
			this.title = context.resources.getString(titleResId)
			return this
		}


		fun withTitle(title: String): Builder {
			this.title = title
			return this
		}


		@SuppressLint("ResourceType")
		fun withTitleItemHeightRes(@DimenRes dimension: Int): Builder {
			this.titleItemHeight = context.resources.getDimensionPixelSize(dimension)
			return this
		}


		@SuppressLint("ResourceType")
		fun withTitleItemHeight(pixelSize: Int): Builder {
			this.titleItemHeight = pixelSize
			return this
		}


		@SuppressLint("ResourceType")
		fun withItemHeightRes(@DimenRes dimension: Int): Builder {
			this.itemHeight = context.resources.getDimensionPixelSize(dimension)
			return this
		}


		@SuppressLint("ResourceType")
		fun withItemHeight(pixelSize: Int): Builder {
			this.itemHeight = pixelSize
			return this
		}


		@SuppressLint("ResourceType")
		fun withImageSizeRes(@DimenRes dimension: Int): Builder {
			this.itemImageSize = context.resources.getDimensionPixelSize(dimension)
			return this
		}


		@SuppressLint("ResourceType")
		fun withImageSize(@DimenRes pixelSize: Int): Builder {
			this.itemImageSize = pixelSize
			return this
		}


		@SuppressLint("ResourceType")
		fun withTitleTextSizeRes(@DimenRes dimension: Int): Builder {
			this.titleTextSize = context.resources.getDimensionPixelSize(dimension)
			return this
		}


		@SuppressLint("ResourceType")
		fun withTitleTextSize(@DimenRes pixelSize: Int): Builder {
			this.titleTextSize = pixelSize
			return this
		}


		@SuppressLint("ResourceType")
		fun withItemTextSizeRes(@DimenRes dimension: Int): Builder {
			this.itemTextSize = context.resources.getDimensionPixelSize(dimension)
			return this
		}


		@SuppressLint("ResourceType")
		fun withItemTextSize(@DimenRes pixelSize: Int): Builder {
			this.itemTextSize = pixelSize
			return this
		}


		@SuppressLint("ResourceType")
		fun withHorizontalSpacingRes(@DimenRes dimension: Int): Builder {
			this.spacingHorizontal = context.resources.getDimensionPixelSize(dimension)
			return this
		}


		@SuppressLint("ResourceType")
		fun withHorizontalSpacing(pixelSize: Int): Builder {
			this.spacingHorizontal = pixelSize
			return this
		}


		@SuppressLint("ResourceType")
		fun withVerticalPaddingRes(@DimenRes dimension: Int): Builder {
			this.paddingVertical = context.resources.getDimensionPixelSize(dimension)
			return this
		}


		@SuppressLint("ResourceType")
		fun withVerticalPadding(pixelSize: Int): Builder {
			this.paddingVertical = pixelSize
			return this
		}


		fun withBackgroundColorRes(@ColorRes color: Int): Builder {
			this.sheetBackgroundColor = ContextCompat.getColor(context, color)
			return this
		}


		fun withBackgroundColor(color: Int): Builder {
			this.sheetBackgroundColor = color
			return this
		}


		fun withItemTouchFeedbackColorRes(@ColorRes color: Int): Builder {
			this.itemTouchFeedbackColor = ContextCompat.getColor(context, color)
			return this
		}


		fun withItemTouchFeedbackColor(color: Int): Builder {
			this.itemTouchFeedbackColor = color
			return this
		}


		fun withTitleColorRes(@ColorRes color: Int): Builder {
			this.sheetTitleColor = ContextCompat.getColor(context, color)
			return this
		}


		fun withTitleColor(color: Int): Builder {
			this.sheetTitleColor = color
			return this
		}


		fun withItemTextColorRes(@ColorRes color: Int): Builder {
			this.itemTextColor = ContextCompat.getColor(context, color)
			return this
		}


		fun withItemTextColor(color: Int): Builder {
			this.itemTextColor = color
			return this
		}


		fun withDividerColorRes(@ColorRes color: Int): Builder {
			this.dividerColor = ContextCompat.getColor(context, color)
			return this
		}


		fun withDividerColor(color: Int): Builder {
			this.dividerColor = color
			return this
		}


		fun build(): HeluBottomButtonSheet {
			val fragment = HeluBottomButtonSheet()
			val arguments = Bundle()

			arguments.putInt(ARGUMENTS_TITLE_ITEM_HEIGHT, this.titleItemHeight)
			arguments.putInt(ARGUMENTS_ITEM_HEIGHT, this.itemHeight)
			arguments.putInt(ARGUMENTS_ITEM_IMAGE_SIZE, this.itemImageSize)
			arguments.putInt(ARGUMENTS_TITLE_TEXT_SIZE, this.titleTextSize)
			arguments.putInt(ARGUMENTS_ITEM_TEXT_SIZE, this.itemTextSize)
			arguments.putInt(ARGUMENTS_SPACING_HORIZONTAL, this.spacingHorizontal)
			arguments.putInt(ARGUMENTS_PADDING_VERTICAL, this.paddingVertical)
			arguments.putInt(ARGUMENTS_ITEM_TOUCH_FEEDBACK_COLOR, this.itemTouchFeedbackColor)
			arguments.putInt(ARGUMENTS_SHEET_BACKGROUND_COLOR, this.sheetBackgroundColor)
			arguments.putInt(ARGUMENTS_SHEET_TITLE_COLOR, this.sheetTitleColor)
			arguments.putInt(ARGUMENTS_ITEM_TEXT_COLOR, this.itemTextColor)
			arguments.putInt(ARGUMENTS_DIVIDER_COLOR, this.dividerColor)
			arguments.putString(ARGUMENTS_TITLE, this.title)

			fragment.arguments = arguments
			return fragment
		}


		private fun convertDpToPx(dp: Int): Int {
			return Math.round(dp * context.resources.displayMetrics.xdpi / DisplayMetrics.DENSITY_DEFAULT)
		}
	}
}
